### PR TITLE
Time span buttons for bar and compare chart

### PR DESCRIPTION
**Pull request by LucasFares**
Adds buttons for common time spans for time and compare charts. Also fixes compare to work properly with typescript changes.

### CODE REVIEW

File: src/client/app/components/UIOptionsComponent.jsx [hunk 1]
```diff
@@ -25,8 +26,12 @@ export default class UIOptionsComponent extends React.Component {
 		this.handleBarDurationChange = this.handleBarDurationChange.bind(this);
 		this.handleBarDurationChangeComplete = this.handleBarDurationChangeComplete.bind(this);
 		this.handleChangeBarStacking = this.handleChangeBarStacking.bind(this);
+		this.handleSpanButton = this.handleSpanButton.bind(this);
+		this.handleCompareSpanButton = this.handleCompareSpanButton.bind(this);
+		this.toggleSlider = this.toggleSlider.bind(this);
 		this.state = {
-			barDuration: this.props.barDuration.asDays()
+			barDuration: this.props.barDuration.asDays(),
+			showsSlider: false
```
Comment (johndiiorio) [inline src/client/app/components/UIOptionsComponent.jsx]: showsSlider -> showSlider

File: src/client/app/components/UIOptionsComponent.jsx [hunk 2]
```diff
@@ -65,10 +70,39 @@ export default class UIOptionsComponent extends React.Component {
...
+		this.props.changeDuration(moment.duration(value, 'days'));
+	}
+
+	handleCompareSpanButton(value) {
+		let compareTimeInterval;
+		let compareDuration;
+		switch (value) {
+			case 'day':
+				compareTimeInterval = new TimeInterval(moment().subtract(2, 'days'), moment()).toString();
+				compareDuration = moment.duration(1, 'hours');
```
Comment (johndiiorio) [inline src/client/app/components/UIOptionsComponent.jsx]: Recommend commenting why you are doing this on an hour basis.

File: src/client/app/components/UIOptionsComponent.jsx [hunk 3]
```diff
@@ -93,14 +132,50 @@ export default class UIOptionsComponent extends React.Component {
 						<div className="checkbox">
 							<label><input type="checkbox" onChange={this.handleChangeBarStacking} checked={this.props.barStacking} />Bar stacking</label>
 						</div>
-						<p style={labelStyle}>Bar chart interval (days):</p>
-						<Slider min={1} max={365} value={this.state.barDuration} onChange={this.handleBarDurationChange} onChangeComplete={this.handleBarDurationChangeComplete} />
+						<p style={labelStyle}>Bar chart interval:</p>
+						<ToggleButtonGroup
```
Comment (johndiiorio) [inline src/client/app/components/UIOptionsComponent.jsx]: Recommend refactoring this. I want this done soon, so it is not a good first issue.

File: src/client/app/containers/CompareChartContainer.js [hunk 4]
```diff
@@ -58,56 +60,60 @@ function mapStateToProps(state, ownProps) {
...
-			// Calculate lastWeek
-			for (let i = 0; i < 7; i++) {
-				lastWeek += readingsData.readings[readingsData.readings.length - (8 + i) - soFar][1];
-			}
+		if (readingsData.readings.length < 7) {
+			soFar = moment().hour();
+		} else if (readingsData.readings.length < 14) {
+			soFar = moment().diff(moment().startOf('week'), 'days');
+		} else {
+			soFar = moment().diff(moment().startOf('week').subtract(21, 'days'), 'days');
```
Comment (johndiiorio) [inline src/client/app/containers/CompareChartContainer.js]: You should comment why this is 21 and not 28.

File: src/client/app/containers/CompareChartContainer.js [hunk 5]
```diff
@@ -58,56 +60,60 @@ function mapStateToProps(state, ownProps) {
...
+		for (let i = readingsData.readings.length - soFar; i < readingsData.readings.length; i++) {
+			current += readingsData.readings[i][1];
+		}
+		// Calculate prev
+		for (let i = 0; i < readingsData.readings.length - soFar; i++) {
+			prev += readingsData.readings[i][1];
+		}
+		// Calculate currentPrev. Have to special case Sunday. Fixing this is a good first issue since we now have hour data.
+		if (soFar === 0) {
+			currentPrev = Math.round((readingsData.readings[0][1] / 24) * moment().hour());
```
Comment (johndiiorio) [inline src/client/app/containers/CompareChartContainer.js]: Recommend not commenting that this is a good first issue, but instead opening a GitHub issue.

File: src/client/app/containers/CompareChartContainer.js [hunk 6]
```diff
@@ -58,56 +60,60 @@ function mapStateToProps(state, ownProps) {
...
-				backgroundColor: color2,
-				hoverBackgroundColor: color2,
-				datalabels: {
-					anchor: 'end',
-					align: 'start',
-				}
+	}
+
+
+	labels.push('Last time interval');
```
Comment (johndiiorio) [inline src/client/app/containers/CompareChartContainer.js]: You need to modify the compare labels to reflect the current time interval.

File: src/client/app/containers/CompareChartContainer.js [hunk 7]
```diff
@@ -145,8 +151,8 @@ function mapStateToProps(state, ownProps) {
 		},
 		title: {
 			display: true,
-			text: delta((-1 + (((currentWeek / currentLastWeek) * lastWeek) / lastWeek))),
-			fontColor: colorize((-1 + (((currentWeek / currentLastWeek) * lastWeek) / lastWeek)))
+			text: delta((-1 + (((current / currentPrev) * prev) / prev))),
```
Comment (johndiiorio) [inline src/client/app/containers/CompareChartContainer.js]: Recommend sticking `(currentWeek / currentLastWeek) * lastWeek) / lastWeek)` in a variable here.

**Comment by johndiiorio** [comment]
You should also update the chart link to include compare duration and compare time interval.

**Comment by johndiiorio** [comment]
You also need the toggle buttons to be selected when the page first loads (to show the selected option).

File: src/client/app/components/UIOptionsComponent.jsx [hunk 8]
```diff
@@ -93,14 +132,50 @@ export default class UIOptionsComponent extends React.Component {
...
+						/>
+						}
 					</div>
+
+				}
+				{this.props.chartToRender === chartTypes.compare &&
+				<div>
+					<ToggleButtonGroup
+						name="timeSpansCompare"
+						value={this.state.compareInterval}
```
Comment (johndiiorio) [inline src/client/app/components/UIOptionsComponent.jsx]: `this.state.compareInterval` is undefined

I believe you need:
`const compareTimeInterval = TimeInterval.fromString(this.props.compareInterval);`
and
`value={compareTimeInterval.endTimestamp.diff(compareTimeInterval.startTimestamp, 'days')}`
as well as modifying the ToggleButton values and the corresponding event handler.

**Comment by johndiiorio** [comment]
This closes #213

**Comment by LucasFares** [comment]
I can't approve things since this is my PR but @johndiiorio's  fixes look good.

**Pull request merged**